### PR TITLE
Support negative strides in `BLAS.gemv!`

### LINKED ISSUE #30765: Bug in matrix times subarray with decreasing index
The code speak for itself
```
A = ones(2,2)
v = [1.0; 2.0]
A * v # [3.0; 3.0]
A * v[2:-1:1] # [3.0; 3.0]
A * view(v, 2:-1:1) # [2.0; 2.0] ????
```

Version info:
```
Julia Version 1.0.3
Commit 099e826241 (2018-12-18 01:34 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```
and also with 1.1.0-rc2
```
Julia Version 1.1.0-rc2.0
Commit 980eda064e (2019-01-13 00:19 UTC)
```

First posted in [discourse](https://discourse.julialang.org/t/bug-in-matrix-times-subarray-with-decreasing-index/19802).

### PR DESCRIPTION
**Pull request by sostock**
Fixes #30765.

This calculates the correct pointers for the BLAS call in the case of vectors with negative stride. I also added an error if `stride(A,2)` doesn’t meet the requirements.

I can also go through the other BLAS functions to see if some of them need to be fixed as well, but I wanted to get feedback on this PR first.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/blas.jl [hunk 1]
```diff
@@ -664,13 +664,19 @@ for (fname, elty) in ((:dgemv_,:Float64),
                 throw(DimensionMismatch("the transpose of A has dimensions $n, $m, X has length $(length(X)) and Y has length $(length(Y))"))
             end
             chkstride1(A)
+            lda = stride(A,2)
+            lda >= max(1, size(A,1)) || error("`stride(A,2)` must be at least `max(1, size(A,1))`")
+            sX = stride(X,1)
+            pX = pointer(X, sX > 0 ? firstindex(X) : lastindex(X))
+            sY = stride(Y,1)
+            pY = pointer(Y, sY > 0 ? firstindex(X) : lastindex(X))
```
Comment (sostock) [inline stdlib/LinearAlgebra/src/blas.jl:672]: Is `pointer` safe to use here or do I need to use `Ref` to prevent the arrays from getting GC’d?

Comment (vtjnash) [reply stdlib/LinearAlgebra/src/blas.jl:672]: you should wrap in `GC.@preserve X Y begin ... end`

Comment (sostock) [reply stdlib/LinearAlgebra/src/blas.jl:672]: Done!

**Comment by ViralBShah** [comment]
Do all BLAS implementations handle this correctly?

**Comment by sostock** [comment]
I tested it with OpenBLAS and MKL (via MKL.jl) on 64-bit Linux, both handle it correctly (at least for the test cases in this PR).

**Comment by ViralBShah** [comment]
@dkarrasch Does this look good? I like the tests and wondering if we can merge.

**Comment by martinholters** [comment]
This broke the case where the output length happened to be zero:
```julia
julia> BLAS.gemv!('N', 1.0, zeros(0, 10), rand(10), 1., zeros(0))
Float64[] # before
ERROR: `stride(A,2)` must be at least `max(1, size(A,1))` # now
```
Was that unsafe (e.g. could cause some memory corruption or the like) to do before? Or is the check too strict, i.e. would `lda >=  size(A,1)` suffice?

**Comment by sostock** [comment]
I haven’t checked whether `lda = size(A,1) = 0` causes any problems, but the references ([netlib](http://www.netlib.org/lapack/explore-html/d7/d15/group__double__blas__level2_gadd421a107a488d524859b4a64c1901a9.html), [MKL](https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/blas-and-sparse-blas-routines/blas-routines/blas-level-2-routines/cblas-gemv.html#cblas-gemv)) state that `lda` must be at least `max(1, m)`. If we make sure that `lda = size(A,1) = 0` works correctly, I’m in favor of relaxing the check to `lda >= size(A,1)`.

**Comment by martinholters** [comment]
Ah, well, with LBT enabling simple switching of the used BLAS library, it's probably better not to expect anything from it not stated in the specs. Still, it's a bit unfortunate this potentially causes breakage. Alternatively, is it safe to set `lda = 1` if `size(A,1) == 0`? (Looking closer, I think this was what effectively used before, rather than `lda==0`.)

**Comment by sostock** [comment]
> Alternatively, is it safe to set `lda = 1` if `size(A,1) == 0`?

I would guess so (I would be surprised if any elements of `A` are accessed if `size(A,1)==0`). I will check and prepare a PR.

**Comment by martinholters** [comment]
_If_ I read the specs correctly, the condition is that `lda*n <= length(A)`, from which it would not be allowed to arbitrarily set `lda=1` for `n>0` just because `m==0`. However, the reference code does have an early exit for `m==0` which would allow this...

**Comment by ViralBShah** [comment]
@martinholters if this is an MKL bug, while we need to address as discussed here, let's also file as a separate issue here and I can report to Intel.

**Comment by martinholters** [comment]
There is no bug in MKL AFAICT. Prior to this change, one could have `size(A,1)==size(Y)==0` and `gemv!` would happily do nothing, whether using MKL or not. And that is also what one would expect when looking at the [Fortran code](http://www.netlib.org/lapack/explore-html/dc/da8/dgemv_8f_source.html#l00155). However, from the docs I think it follows that this is actually not allowed. And this PR replaces the "doesn't do anything" with a "throws an error". So I guess the question boils down to: May we expect other BLAS libraries to behave like the Fortran code, or could they instead require what the original documentation said? As an important data point, MKL seems to be inline with the Fortran code in this regard.

**Pull request merged**